R/getShortTermChange.R

```diff
@@ -0,0 +1,146 @@
+
+#' @rdname getShortTermChange
+#' @export
+#' @importFrom dplyr arrange as_tibble summarize "%>%"
+#' @importFrom ggplot2 ggplot aes
+#' @importFrom ggrepel geom_text_repel
+#' @importFrom mia rarefyAssay transformAssay
+#' @importFrom SummarizedExperiment colData
+setMethod("getShortTermChange", signature = c(x = "SummarizedExperiment"),
+    function(x, assay.type = "counts", rarefy = FALSE, compositional = FALSE, 
+        depth = min(assay(x, assay.type)), ...){
+        ############################## Input check #############################
+        # Check validity of object
+        if(nrow(x) == 0L){
+            stop("No data available in `x` ('x' has nrow(x) == 0L.)",
+                 call. = FALSE)
+        }
+        # Check assay.type
+        .check_assay_present(assay.type, x)
+        if(!.is_a_bool(rarefy)){
+            stop("'rarefy' must be TRUE or FALSE.", call. = FALSE)
+        }
+        if(!.is_a_bool(compositional)){
+            stop("'compositional' must be TRUE or FALSE.", call. = FALSE)
+        }
+        # Ensure that the provided depth is valid
+        if ( !is.null(depth) && depth > min(assay(x, assay.type), na.rm = TRUE) ) {
+            stop("Depth cannot be greater than the minimum number of counts in your data", call. = FALSE)
+        
+        }
+        ########################### Growth Metrics ############################
+        grwt <- .calculate_growth_metrics(x, assay.type = assay.type, 
+                                          rarefy = rarefy, 
+                                          compositional = compositional, 
+                                          depth = depth, ...)
+        # Clean and format growth matrics
+        grs.all <- .clean_growth_metrics(grwt, ...)
+        return(grs.all)
+    }
+)
+# wrapper to calculate growth matrix
+.calculate_growth_metrics <- function(x, assay.type, time.col = NULL, 
+                                      rarefy, compositional, depth, ...) {
+    ############################ Data Preparation ##############################
+    # Initialize the filtered object based on rarefy and compositional arguments
+    if (rarefy == TRUE && compositional == FALSE) {
+        message("rarefy is set to TRUE, calculating short term change using counts")
+        x <- rarefyAssay(x, assay.type = assay.type, depth = depth)
+        assay.type <- "subsampled"
+    } else if (rarefy == FALSE && compositional == FALSE) {
+        message("rarefy is set to FALSE, compositional==FALSE, using raw counts")
+        x <- x
+    } else if (rarefy == FALSE && compositional == TRUE) {
+        message("rarefy is set to FALSE, compositional==TRUE, using relative abundances")
+        x <- transformAssay(x, method = "relabundance", assay.type = assay.type)
+        assay.type <- "relabundance"
+    } else if (rarefy == TRUE && compositional == TRUE) {
+        stop("Both rarefy and compositional cannot be TRUE simultaneously", call. = FALSE)
+    }
```

Comment:
I think these should be done outside the function

Reply:
Rarification is part of calculating beta diversity (ie. dissimilarity), whereever that happens.

Reply:
But this could be done outside the function to avoid duplicated calculations. We have also other functions where relative abundances are meant to be used, but user is advised to do it before running the function.

Reply:
Yes I agree. I meant that if beta diversity calculation is in some cases part of the procedure within the function then the rarefaction arguments should be passed rather than reimplemented in the function. But now I realize that we do not have rarified version of the beta diversity calculation - that was problematic to implement because rarefaction can be followed by transformation before dissimilarity calculation and in general we do not like mix up these too much.. although I still may think that perhaps it would be rather handy to support rarefaction in `getDissimilarity` but that's another question..

Reply:
`getDissimilarity` actually nowadays support rarefaction via using `vegan::avgdist`

https://microbiome.github.io/mia/reference/getDissimilarity.html